File: frontend/src/lib/api/scope3.ts
import apiClient from "./client";

export interface MaterialBreakdownItem {
  material_id: string;
  material_name: string;
  weight_pct: number;
  co2e_per_kg: number | null;
  contribution_kg_co2e: number | null;
}

export interface ProductCarbonFootprint {
  id: string;
  organization_id: string;
  product_id: string;
  reporting_year: number;
  pcf_kg_co2e_per_unit: number | null;
  pcf_source: string;
  bom_materials_total: number;
  bom_materials_with_lca: number;
  weight_coverage_pct: number | null;
  material_breakdown: MaterialBreakdownItem[];
  calc_version: string;
  calculated_at: string;
  calculated_by: string | null;
  notes: string | null;
}

export interface ProductCarbonFootprintListResponse {
  items: ProductCarbonFootprint[];
  total: number;
}

export interface Scope3Inventory {
  id: string;
  organization_id: string;
  reporting_year: number;
  total_pcf_kg_co2e: number;
  total_pcf_tco2e: number;
  products_included: number;
  products_with_full_lca: number;
  products_with_partial_lca: number;
  products_without_lca: number;
  top_contributors: Array<{
    product_id: string;
    pcf_kg_co2e: number;
    weight_coverage_pct: number | null;
  }>;
  calc_version: string;
  calculated_at: string;
  calculated_by: string | null;
}

export interface Scope3InventoryListResponse {
  items: Scope3Inventory[];
  total: number;
}

export interface Scope3OrgSummary {
  organization_id: string;
  reporting_year: number | null;
  total_products_with_pcf: number;
  total_pcf_kg_co2e: number;
  total_pcf_tco2e: number;
  avg_pcf_kg_co2e_per_product: number | null;
  lca_coverage_pct: number | null;
}

export async function calculatePCF(
  productId: string,
  reportingYear: number,
  notes?: string
): Promise<ProductCarbonFootprint> {
  const { data } = await apiClient.post<ProductCarbonFootprint>(
    `/scope3/pcf/${productId}`,
    { reporting_year: reportingYear, notes: notes ?? null }
  );
  return data;
}

export async function listProductPCFs(
  productId: string,
  limit = 20
): Promise<ProductCarbonFootprintListResponse> {
  const { data } = await apiClient.get<ProductCarbonFootprintListResponse>(
    `/scope3/pcf/${productId}`,
    { params: { limit } }
  );
  return data;
}

export async function listOrgPCFs(params?: {
  reporting_year?: number;
  limit?: number;
}): Promise<ProductCarbonFootprintListResponse> {
  const { data } = await apiClient.get<ProductCarbonFootprintListResponse>(
    "/scope3/pcf",
    { params }
  );
  return data;
}

export async function computeInventory(year: number): Promise<Scope3Inventory> {
  const { data } = await apiClient.post<Scope3Inventory>(`/scope3/inventory/${year}`);
  return data;
}

export async function listInventories(): Promise<Scope3InventoryListResponse> {
  const { data } = await apiClient.get<Scope3InventoryListResponse>("/scope3/inventory");
  return data;
}

export async function getInventory(year: number): Promise<Scope3Inventory> {
  const { data } = await apiClient.get<Scope3Inventory>(`/scope3/inventory/${year}`);
  return data;
}

export async function getScope3Summary(reportingYear?: number): Promise<Scope3OrgSummary> {
  const { data } = await apiClient.get<Scope3OrgSummary>("/scope3/summary", {
    params: reportingYear ? { reporting_year: reportingYear } : undefined,
  });
  return data;
}

File: backend/interfaces/api/schemas/scope3.py
"""Pydantic schemas — M8 Scope 3 Supply Chain Carbon Inventory."""
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict

from infrastructure.persistence.models.scope3 import (
    ProductCarbonFootprintModel,
    Scope3InventoryModel,
)


class MaterialBreakdownItem(BaseModel):
    material_id: str
    material_name: str
    weight_pct: float
    co2e_per_kg: float | None
    contribution_kg_co2e: float | None


class ProductCarbonFootprintResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    organization_id: str
    product_id: str
    reporting_year: int
    pcf_kg_co2e_per_unit: float | None
    pcf_source: str
    bom_materials_total: int
    bom_materials_with_lca: int
    weight_coverage_pct: float | None
    material_breakdown: list[dict]
    calc_version: str
    calculated_at: datetime
    calculated_by: str | None
    notes: str | None

    @classmethod
    def from_model(cls, m: ProductCarbonFootprintModel) -> "ProductCarbonFootprintResponse":
        return cls.model_validate(m)


class ProductCarbonFootprintListResponse(BaseModel):
    items: list[ProductCarbonFootprintResponse]
    total: int


class PCFCalculateRequest(BaseModel):
    reporting_year: int
    notes: str | None = None


class TopContributor(BaseModel):
    product_id: str
    pcf_kg_co2e: float
    weight_coverage_pct: float | None


class Scope3InventoryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    organization_id: str
    reporting_year: int
    total_pcf_kg_co2e: float
    total_pcf_tco2e: float
    products_included: int
    products_with_full_lca: int
    products_with_partial_lca: int
    products_without_lca: int
    top_contributors: list[dict]
    calc_version: str
    calculated_at: datetime
    calculated_by: str | None

    @classmethod
    def from_model(cls, m: Scope3InventoryModel) -> "Scope3InventoryResponse":
        return cls.model_validate(m)


class Scope3InventoryListResponse(BaseModel):
    items: list[Scope3InventoryResponse]
    total: int


class Scope3OrgSummaryResponse(BaseModel):
    organization_id: str
    reporting_year: int | None
    total_products_with_pcf: int
    total_pcf_kg_co2e: float
    total_pcf_tco2e: float
    avg_pcf_kg_co2e_per_product: float | None
    lca_coverage_pct: float | None
